test/e2e/cloud.e2e.js

```diff
@@ -74,14 +74,121 @@ describe('Cloud Commands [@device]', () => {
+
+				expect(stdout).to.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+
+			it('Filters out non-matching devices', async () => {
+				// Likely platforms being used with e2e testing
+				const testPlatforms = [
+					'photon',
+					'electron',
+					'argon',
+					'boron'
+				];
+
+				// Derive platform name NOT matching those claimed by E2E test profile
+				// (i.e. find single, valid inversion of DEVICE_PLATFORM_NAME)
+				const nonE2EDevicePlatform = testPlatforms.filter((platform) => {
+					return platform !== DEVICE_PLATFORM_NAME;
+				})[0];
+
+				args.push(nonE2EDevicePlatform);
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.not.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+		});
+
+		describe('Filter by Device Name', () => {
+			it('Shows matching devices', async () => {
+				args.push(DEVICE_NAME);
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+
+			it('Filters out non-matching devices', async () => {
+				args.push('non-existent-device-name');
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.not.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+		});
+
+		describe('Filter by Device ID', () => {
+			it('Shows matching devices', async () => {
+				args.push(DEVICE_ID);
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+
+			it('Filters out non-matching devices', async () => {
+				args.push('non-existent-device-id');
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.not.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+		});
+
+		describe('Filter by online', () => {
+			it('Shows online devices', async () => {
+				args.push('online');
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+		});
+
+		describe('Filter by offline', () => {
+			it('Does not show online devices', async () => {
+				args.push('offline');
+				const { stdout, stderr, exitCode } = await cli.run(args);
+
+				expect(stdout).to.not.include(`${DEVICE_NAME} [${DEVICE_ID}] (${platform})`);
+				expect(stderr).to.equal('');
+				expect(exitCode).to.equal(0);
+			});
+		});
```

Comment:
nice :clap::clap::clap::+1:

i'll probably tune the naming / grouping of these a bit to better align with what i've been doing elsewhere ([example](https://github.com/particle-iot/particle-cli/blob/42091eee5bcb6779c163a30434ec9d9d83b14b0a/test/e2e/keys.e2e.js#L277-L320)) but i don't think that needs to block merging. fwiw, it matters b/c at some point i think we're going to need to be more granular in our categorization of tests (`@device` tag and related) and having them clearly grouped by command makes that more practical (setup / teardown is super-expensive atm).

Reply:
Ah, yeah... Those tests are slow and the categorization is a big reason why.

Honestly, the categorization of those tests bit me here. The `Names a device` test doesn't clean itself up and expects to be ran last. Adding my `describe` block broke that assumption and required me to call `await cli.revertDeviceName();`. I'm sure there's a better way to handle that cleanup...